src/nwb/RegisteredType.hpp

```diff
@@ -0,0 +1,327 @@
+ * @param TYPENAME The name of the type (usually the class name).
+ */
+#define REGISTER_SUBCLASS_WITH_TYPENAME(T, NAMESPACE, TYPENAME) \
+  static bool registerSubclass() \
+  { \
+    AQNWB::NWB::RegisteredType::registerSubclass( \
+        NAMESPACE "::" #T, \
+        [](const std::string& path, std::shared_ptr<IO::BaseIO> io) \
+            -> std::unique_ptr<AQNWB::NWB::RegisteredType> \
+        { return std::make_unique<T>(path, io); }, \
+        TYPENAME, \
+        NAMESPACE); \
+    return true; \
+  } \
+  static bool registered_; \
+  virtual std::string getTypeName() const override \
+  { \
+    return TYPENAME; \
+  } \
+  virtual std::string getNamespace() const override \
+  { \
+    return NAMESPACE; \
+  }
+
+/**
+ * @brief Macro to register a subclass with the RegisteredType class registry.
+ *
+ * This macro is a convenience wrapper around the main REGISTER_SUBCLASS macro,
+ * providing a default value for TYPENAME.
+ *
+ * @param T The subclass type to register. The name must match the type in the
+ * schema.
+ * @param NAMESPACE The namespace of the subclass type in the format schema
+ */
+#define REGISTER_SUBCLASS(T, NAMESPACE) \
+  REGISTER_SUBCLASS_WITH_TYPENAME(T, NAMESPACE, #T)
+
+/**
+ * @brief Macro to initialize the static member `registered_` to trigger
+ * registration.
+ *
+ * This macro ensures that the registration of the subclass occurs when the
+ * program starts.
+ *
+ * @param T The subclass type to register.
+ */
+#define REGISTER_SUBCLASS_IMPL(T) bool T::registered_ = T::registerSubclass();
+
+/**
+ * @brief Defines a lazy-loaded field accessor function.
+ *
+ * This macro generates a function that returns a lazy-loaded wrapper for a
+ * field.
+ *
+ * \note
+ * The Doxyfile.in defines a simplified expansion of this function
+ * for generating the documentation for the autogenerated function.
+ * This means: 1) When updating the macro here, we also need to ensure
+ * that the expansion in the Doxyfile.in is still accurate and 2) the
+ * docstring that is defined by the macro here is not being used by
+ * Doxygen but the version generated by its on PREDEFINED expansion.
+ *
+ * @param name The name of the function to generate.
+ * @param storageObjectType The type of storage object (Attribute or Dataset).
+ * @param default_type The default type of the field.
+ * @param fieldPath The path to the field.
+ * @param description A detailed description of the field.
+ */
+#define DEFINE_FIELD( \
+    name, storageObjectType, default_type, fieldPath, description) \
+  /** \
+   * @brief Returns a lazy-loaded wrapper for the ##name field. \
+   * \
+   * @tparam VTYPE The type of the field (default: ##default_type) \
+   * @return A unique pointer to a ReadDataWrapper for the field \
+   * \
+   * description \
+   */ \
+  template<typename VTYPE = default_type> \
+  inline std::unique_ptr<IO::ReadDataWrapper<storageObjectType, VTYPE>> name() \
```

Comment:
```suggestion
  inline virtual std::unique_ptr<IO::ReadDataWrapper<storageObjectType, VTYPE>> name() \
```

We should add the ``virtual`` keyword here to ensure when overriding with `DEFINE_FIELD` in subclasses, that the function resolution works as expected. I am not sure how much sense it makes to keep `inline` as well or whether that should be removed.

Reply:
Actually, that is not going to work, since the function is also a template function and virtual and templates don't like each other. As it is currently, when`DEFINE_FIELD` is specified again in a derived class for the same field, then the function will be hidden instead of overriden . I.e., if we call the function on the derived class everything works as expected. However, if we cast the Derived pointer to the base class, then we will call the function as defined on the base class, because it is not virtual. I think in most cases this should not be a problem, but is something we should be aware of, and is worth documenting. 

A possible workaround would be to have `DEFINE_FIELD` generate two member functions. A templated function (as it is right now) and then a second virtual function that calls the templated function with the `default_type`

```
 virtual std::unique_ptr<IO::ReadDataWrapper<storageObjectType, default_type>> name##_virtual() const \
  { \
    return name<default_type>(); \
  }
```

However, I am not sure this is necessary and I think could lead to more confusion. Maybe we can just make an issue for this for now just have a record for this.

Reply:
I created #117 for this. I think it is fine to address this later. For now, I added the following note to the docs:

https://github.com/NeurodataWithoutBorders/aqnwb/blob/7b88122ce7a1855f5c640d8235cfe18ab3355c01/docs/pages/devdocs/registered_types.dox#L82-L87